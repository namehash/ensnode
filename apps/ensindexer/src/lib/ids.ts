import config from "@/config";
import { type LabelHash, type Node } from "@ensnode/ensnode-sdk";
import type { Address } from "viem";

/**
 * Makes a unique, chain-scoped resolver ID.
 * In Subgraph-compatibility mode, no chainId prefix is used (subgraph-compat).
 *
 * @example Subgraph-compat: `${address}-${node}`
 * @example Otherwise (chain-scoped): `${chainId}-${address}-${node}`
 *
 * @param chainId the chain ID
 * @param address the resolver contract address
 * @param node the ENS node
 * @returns a unique resolver ID
 */

export const makeResolverId = (chainId: number, address: Address, node: Node) =>
  [
    // null out chainId prefix iff subgraph-compat, otherwise include for chain-scoping
    config.isSubgraphCompatible ? null : chainId,
    // NOTE: subgraph uses lowercase address here, viem provides us checksummed, so we lowercase it
    address.toLowerCase(),
    node,
  ]
    .filter(Boolean)
    .join("-");

/**
 * Makes a unique, chain-scoped event ID.
 * In Subgraph-compatibility mode, no chainId prefix is used (subgraph-compat).
 *
 * @example Subgraph-compat: `${blockNumber}-${logIndex}(-${transferIndex})`
 * @example Otherwise (chain-scoped): `${chainId}-${blockNumber}-${logIndex}(-${transferIndex})`
 *
 * @param chainId
 * @param blockNumber
 * @param logIndex
 * @param transferIndex
 * @returns
 */
export const makeEventId = (
  chainId: number,
  blockNumber: bigint,
  logIndex: number,
  transferIndex?: number,
) =>
  [
    // null out chainId prefix iff subgraph plugin, otherwise include for chain-scoping
    config.isSubgraphCompatible ? null : chainId,
    blockNumber.toString(),
    logIndex.toString(),
    transferIndex?.toString(),
  ]
    .filter(Boolean)
    .join("-");

/**
 * Makes a cross-registrar unique registration ID.
 *
 * The ENS Subgraph only indexes Registration entities for a single registrar: the registrar for
 * direct subnames of ".eth". It uses the labelHash of each registered ".eth" subname to
 * form a unique Registration id.
 *
 * Because ENSIndexer supports indexing multiple Registrar contracts via plugins
 * (currently using the shared handlers modelled after the Subgraph's indexing logic), however,
 * additional Registration entities may be created. A unique ID other than labelHash is necessary,
 * otherwise Registration entities for the same label would collide.
 *
 * To avoid collisions, if the caller identifies as the subgraph plugin, we use the Domain's `labelHash`
 * (subgraph compat). Otherwise, for any other plugin, we use the Domain's `node`, which is
 * globally unique within ENS.
 *
 * We knowingly mix `labelHash` (labelhash) and `node` (namehash) values as registration ids:
 * both result in keccak256 hashes (the odds of a collision being practically zero) and are derived
 * differently â€” the result of `namehash` will never (practically zero) collide with the result of
 * `labelhash` because `namehash` always includes the recursive hashing of the root node.
 *
 * For the "v1" of ENSIndexer (at a minimum) we want to preserve exact backwards compatibility with
 * Registration IDs issued by the ENS Subgraph. In the future we may relax exact subgraph backwards
 * compatibility and use `node` for all Registration IDs.
 *
 * @param labelHash the labelHash of the name that was registered
 * @param node the node of the name that was registered
 * @returns a unique registration id
 */
export const makeRegistrationId = (labelHash: LabelHash, node: Node) => {
  if (config.isSubgraphCompatible) return labelHash;
  return node;
};

/**
 * Makes a unique ID for any resolver record entity that is keyed beyond Node
 * (i.e. text, address records).
 *
 * See comment in packages/ensnode-schema/src/subgraph.schema.ts for additional context.
 *
 * @example
 * ```ts
 * // For address records in a Resolver, use coinType as key
 * makeKeyedResolverRecordId(resolverId, coinType.toString())
 * // => "0x123...-60" // for ETH (coinType 60)
 *
 * // For text records in a Resolver, use the text key
 * makeKeyedResolverRecordId(resolverId, "avatar")
 * // => "0x123...-avatar"
 * ```
 *
 * @param resolverId the id of the resolver entity
 * @param key the unique id of the resolver record within a resolverId
 * @returns a unique resolver record entity id
 */
export const makeKeyedResolverRecordId = (resolverId: string, key: string) =>
  [resolverId, key].join("-");
