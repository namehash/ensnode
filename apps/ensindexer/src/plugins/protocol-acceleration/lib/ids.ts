import { ChainId, CoinType, Node, coinTypeReverseLabel } from "@ensnode/ensnode-sdk";
import { Address } from "viem";

/**
 * Makes a unique ID for any resolver record entity that is keyed beyond Node
 * (i.e. text, address records).
 *
 * See comment in packages/ensnode-schema/src/subgraph.schema.ts for additional context.
 *
 * @example
 * ```ts
 * // For address records in a Resolver, use coinType as key
 * makeKeyedResolverRecordId(resolverId, coinType.toString())
 * // => "0x123...-60" // for ETH (coinType 60)
 *
 * // For text records in a Resolver, use the text key
 * makeKeyedResolverRecordId(resolverId, "avatar")
 * // => "0x123...-avatar"
 * ```
 *
 * @param resolverId the id of the resolver entity
 * @param key the unique id of the resolver record within a resolverId
 * @returns a unique resolver record entity id
 */
export const makeKeyedResolverRecordId = (resolverId: string, key: string) =>
  [resolverId, key].join("-");

/**
 * Makes a unique ID for a Node<->Resolver relation on a given chain. NodeResolverRelations are
 * unique by (chainId, node).
 *
 * @see packages/ensnode-schema/src/protocol-acceleration.schema.ts
 *
 * @example `${chainId}-${node}`
 *
 * @param chainId the chain ID
 * @param node the Node
 * @returns a unique Node-Resolver relation ID
 */
export const makeNodeResolverRelationId = (chainId: ChainId, node: Node) =>
  [chainId, node].join("-");

/**
 * Makes a unique ID for a primary name record, keyed by (address, coinType).
 *
 * @param address the address for which the primary name is set
 * @param coinType the coin type
 * @returns a unique primary name id
 */
export const makePrimaryNameId = (address: Address, coinType: CoinType) =>
  [address, coinTypeReverseLabel(coinType)].join("-");

/**
 * Makes a unique Resolver Records ID.
 *
 * @example `${chainId}-${address}-${node}`
 *
 * @param chainId the chain ID
 * @param address the resolver contract address
 * @param node the ENS Node
 * @returns a unique Resolver Records ID
 */

export const makeResolverRecordsId = (chainId: number, address: Address, node: Node) =>
  [chainId, address, node].join("-");
