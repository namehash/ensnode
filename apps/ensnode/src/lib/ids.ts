import type { Address, Hex } from "viem";
import type { Labelhash, Node, OwnedName } from "./types";

// NOTE: subgraph uses lowercase address here, viem provides us checksummed, so we lowercase it
export const makeResolverId = (address: Address, node: Hex) =>
  [address.toLowerCase(), node].join("-");

// https://github.com/ensdomains/ens-subgraph/blob/master/src/utils.ts#L5
// produces `blocknumber-logIndex` or `blocknumber-logindex-transferindex`
export const makeEventId = (blockNumber: bigint, logIndex: number, transferIndex?: number) =>
  [blockNumber.toString(), logIndex.toString(), transferIndex?.toString()]
    .filter(Boolean)
    .join("-");

/**
 * Makes a cross-chain unique registration ID.
 *
 * A Registration record's id is historically the labelhash of the label
 * directly under the name the Registrar manages (i.e. for a registration
 * of test.eth, a Registration's id is labelhash('test')). Because the original
 * subgraph implementation didn't consider indexing any Registrars besides
 * the .eth Registrar, this leaves no room in the namespace for Registration
 * records from other Registrars (like the base.eth and linea.eth Registrars).
 * To account for this, while preserving backwards compatibility, Registration
 * records created for the .eth Registrar use label as id and those created by
 * any other Registrar use node (i.e. namehash(test.base.eth) to avoid
 * collisions that would otherwise occur.
 *
 * @param ownedName the ownedName of the ENSNode plugin that is processing the registration event
 * @param label registration's label
 * @param node registration's node
 */
export const makeRegistrationId = (ownedName: OwnedName, label: Labelhash, node: Node) =>
  ownedName === "eth" ? label : node;
