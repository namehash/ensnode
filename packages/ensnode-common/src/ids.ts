import type { Address, Hex } from "viem";

// NOTE: subgraph uses lowercase address here, viem provides us checksummed, so we lowercase it
export const makeResolverId = (address: Address, node: Hex) =>
  [address.toLowerCase(), node].join("-");

// https://github.com/ensdomains/ens-subgraph/blob/master/src/utils.ts#L5
// produces `blocknumber-logIndex` or `blocknumber-logindex-transferindex`
export const makeEventId = (blockNumber: bigint, logIndex: number, transferIndex?: number) =>
  [blockNumber.toString(), logIndex.toString(), transferIndex?.toString()]
    .filter(Boolean)
    .join("-");
