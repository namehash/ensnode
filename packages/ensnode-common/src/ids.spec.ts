import { namehash, zeroAddress } from "viem";
import { describe, expect, it } from "vitest";
import { makeEventId, makeResolverId } from "./ids";

describe("ids", () => {
  describe("makeResolverId", () => {
    it("should match snapshot", () => {
      expect(makeResolverId(zeroAddress, namehash("vitalik.eth"))).toMatchSnapshot();
    });
  });

  describe("makeEventId", () => {
    it("should match snapshot", () => {
      expect(makeEventId(123n, 456, 1)).toMatchSnapshot();
      expect(makeEventId(123n, 456)).toMatchSnapshot();
    });
  });
});
