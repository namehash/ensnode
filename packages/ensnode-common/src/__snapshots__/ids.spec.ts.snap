// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ids > makeEventId > should match snapshot 1`] = `"123-456-1"`;

exports[`ids > makeEventId > should match snapshot 2`] = `"123-456"`;

exports[`ids > makeResolverId > should match snapshot 1`] = `"0x0000000000000000000000000000000000000000-0xee6c4522aab0003e8d14cd40a6af439055fd2577951148c14b6cea9a53475835"`;
